### ISSUE TITLE
vkSetEvent validation error Code 66: Cannot call vkSetEvent() on event XX that is already in use by a command buffer

### ISSUE THREAD

**Issue by aircow33** [KhronosGroup/Vulkan-LoaderAndValidationLayers#807]
My code is pretty straightforward

it basically codes this

vkSetEvent(Vulkan->getDevice(), event);
vkCmdResetEvent(mCurrentCmdBuffer, event, VK_PIPELINE_STAGE_ALL_GRAPHICS_BIT);
....
....
and then later

if (vkGetEventStatus(mVulkan->getDevice(), event) == VK_EVENT_RESET)
{
     //vkDeviceWaitIdle(mVulkan->getDevice());   if i uncomment this line, no validation errors
     vkSetEvent(mVulkan->getDevice(), event);  //blows up here
     vkCmdResetEvent(mCurrentCmdBuffer, event, VK_PIPELINE_STAGE_ALL_GRAPHICS_BIT);
}

Am I missing something?  if event has been reset, why can i not set the event on the cpu side?

**Comment by chrisforbes** [comment]
Core validation currently doesn't even hook `vkGetEventStatus`, so no surprise that it's failing to reason about your case. This is going to need some thought. @lentinem?

**Comment by lentinem** [comment]
Looks like we assume that the command buffer has to have been completed before allowing set but yes vkGetEventStatus on returned reset should also be a case where this is allowed. Shouldn't take too much effort to add that to geteventstatus.

**Comment by lentinem** [comment]
I noticed something odd with your code though. If you are calling getstatus on a command buffer that has not completed. How are you calling resetevent on that same command buffer? Even if we did modify geteventstatus I would expect the exact same error to be thrown at that point.

**Comment by lentinem** [comment]
On further inspection even with the command I don't think this is valid. There is no guarantee that the stage is still in RESET when calling vkSetEvent as the state might have been changed by the in flight command buffer so I think you are going to have to call wait on the fence associated with that command buffer for this to work.

**Comment by aircow33** [comment]
Are you saying i cannot do a vkGetEventStatus check until all commands before the fence have been executed?   I am doing a vkGetEventStatus to check if the event state is in reset.   The command i submit to the GPU is to put it in the reset state.  The vkGetEventStatus doesnt take commandbuffer as a parameter.  It should just return back the state of the event.   

Events, providing a finer-grained synchronization primitive which can be signaled at command level granularity by both device and host, and can be waited upon by either.

The fact that vkGetEventStatus is returning reset status means the command has already been executed.  If it hasnt been executed then it would still be in the set state.

**Comment by lentinem** [comment]
Checking for the status is ok. In this particular case you are only issuing one command in the buffer (reset) so there is enough knowledge to know if that command has completed via getstatus (although in general this may not be true). The main issue though is you cannot modify a command buffer that is in flight which you appear to be doing with the second call to vkcmdresetevent.

**Comment by aircow33** [comment]
yeah the second call is inside the if case....so it means that the command has already be executed no?

**Comment by lentinem** [comment]
The command is already executed but the command buffer may not have completed.

**Comment by aircow33** [comment]
i see, yeah sorry mCurrentCmdBuffer is actually a different commandbuffer, we are creating a commandbuffer for each presentation image.   sorry for the confusion.

**Comment by lentinem** [comment]
Yes it does; it's not an easy fix as it's a very specific case that this works in which you present here so it might take some time.
